From 0000000000000000000000000000000000000002 Mon Sep 17 00:00:00 2001
From: QEMU Haiku Builder <builder@localhost>
Date: Thu, 31 Jul 2026 11:00:00 +0900
Subject: Haiku: accept mapped pthread condition timeout status

When posix_error_mapper is linked on Haiku x86, pthread_cond_timedwait()
returns the positive counterpart of Haiku's negative ETIMEDOUT constant.
Treat both representations as a normal timeout.

diff --git a/util/qemu-thread-posix.c b/util/qemu-thread-posix.c
index 5c0d304..2222222 100644
--- a/util/qemu-thread-posix.c
+++ b/util/qemu-thread-posix.c
@@ -238,10 +238,15 @@ qemu_cond_timedwait_ts(QemuCond *cond, QemuMutex *mutex, struct timespec *ts,
     trace_qemu_mutex_unlock(mutex, file, line);
     err = pthread_cond_timedwait(&cond->cond, &mutex->lock, ts);
     trace_qemu_mutex_locked(mutex, file, line);
-    if (err && err != ETIMEDOUT) {
+#ifdef __HAIKU__
+    bool timed_out = err == ETIMEDOUT || err == -ETIMEDOUT;
+#else
+    bool timed_out = err == ETIMEDOUT;
+#endif
+    if (err && !timed_out) {
         error_exit(err, __func__);
     }
-    return err != ETIMEDOUT;
+    return !timed_out;
 }
 
 bool qemu_cond_timedwait_impl(QemuCond *cond, QemuMutex *mutex, int ms,
